core/services/offchainreporting2/delegate.go

```diff
@@ -114,92 +123,72 @@ func (d Delegate) ServicesForSpec(jobSpec job.Job) (services []job.Service, err
-				jobSpec,
-				*jobSpec.PipelineSpec,
-				loggerWith,
-				runResults,
-			),
-			JuelsPerFeeCoinDataSource: ocrcommon.NewInMemoryDataSource(d.pipelineRunner, jobSpec, juelsPerFeeCoinPipelineSpec, loggerWith),
-			ReportCodec:               ocr2Provider.ReportCodec(),
-			Logger:                    ocrLogger,
-		}
-
-		jobSpec.PipelineSpec.JobName = jobSpec.Name.ValueOrZero()
-		jobSpec.PipelineSpec.JobID = jobSpec.ID
-		oracle, err := libocr2.NewOracle(libocr2.OracleArgs{
-			BinaryNetworkEndpointFactory: peerWrapper.Peer2,
-			V2Bootstrappers:              bootstrapPeers,
-			ContractTransmitter:          ocr2Provider.ContractTransmitter(),
-			ContractConfigTracker:        tracker,
-			Database:                     ocrDB,
-			LocalConfig:                  lc,
-			Logger:                       ocrLogger,
-			MonitoringEndpoint:           d.monitoringEndpointGen.GenMonitoringEndpoint(spec.ContractID),
-			OffchainConfigDigester:       offchainConfigDigester,
-			OffchainKeyring:              kb,
-			OnchainKeyring:               kb,
-			ReportingPluginFactory:       numericalMedianFactory,
-		})
-		if err != nil {
-			return nil, errors.Wrap(err, "error calling NewOracle")
-		}
-		services = append(services, oracle)
-
-		// RunResultSaver needs to be started first so its available
-		// to read odb writes. It is stopped last after the Oracle is shut down
-		// so no further runs are enqueued and we can drain the queue.
-		services = append([]job.Service{ocrcommon.NewResultRunSaver(
-			runResults,
-			d.pipelineRunner,
-			make(chan struct{}),
+	bootstrapPeers, err := ocrcommon.GetValidatedBootstrapPeers(spec.P2PBootstrapPeers, peerWrapper.Config().P2PV2Bootstrappers())
+	if err != nil {
+		return nil, err
+	}
+	d.lggr.Debugw("Using bootstrap peers", "peers", bootstrapPeers)
+	// Fetch the specified OCR2 key bundle
+	var kbID string
+	if spec.OCRKeyBundleID.Valid {
+		kbID = spec.OCRKeyBundleID.String
+	} else if kbID, err = d.cfg.OCR2KeyBundleID(); err != nil {
+		return nil, err
+	}
+	kb, err := d.ks.Get(kbID)
+	if err != nil {
+		return nil, err
+	}
+
+	runResults := make(chan pipeline.Run, d.cfg.JobPipelineResultWriteQueueDepth())
+	juelsPerFeeCoinPipelineSpec := pipeline.Spec{
+		ID:           jobSpec.ID,
+		DotDagSource: spec.JuelsPerFeeCoinPipeline,
+		CreatedAt:    time.Now(),
+	}
+	numericalMedianFactory := median.NumericalMedianFactory{
+		ContractTransmitter: ocr2Provider.MedianContract(),
+		DataSource: ocrcommon.NewDataSourceV2(d.pipelineRunner,
+			jobSpec,
+			*jobSpec.PipelineSpec,
 			loggerWith,
-		)}, services...)
+			runResults,
+		),
+		JuelsPerFeeCoinDataSource: ocrcommon.NewInMemoryDataSource(d.pipelineRunner, jobSpec, juelsPerFeeCoinPipelineSpec, loggerWith),
+		ReportCodec:               ocr2Provider.ReportCodec(),
+		Logger:                    ocrLogger,
+	}
+
+	jobSpec.PipelineSpec.JobName = jobSpec.Name.ValueOrZero()
+	jobSpec.PipelineSpec.JobID = jobSpec.ID
```

Comment:
This feels hacky. Why does this need to be set here?

Reply:
ah this is required for prometheus metrics https://github.com/smartcontractkit/chainlink/pull/4147/files. I think it can be made slightly less hacky if we populated the job spec fields in the pipeline spec in the job spawner at load time. Separately however I think there's a current bug in that we need to set this _before_ we call ocrcommon.NewDataSourceV2 or it will not be present there

Reply:
IMO if this is broken, fixing it should be in the scope of this PR since its copypasta code.

Ideally I'd like to see all this stuff abstracted into ocr common